### ISSUE TITLE
Random deviates from standard algortihm

### ISSUE THREAD

**Issue by terrajobst** [dotnet/runtime#23198]
A customer filed the following bug report on [Connect](https://connect.microsoft.com/VisualStudio/feedback/details/634761/system-random-serious-bug):

> While investigating the period of various random number generators, I found a serious bug in Microsoft .NET System.Random implementation.
>
> [Microsoft Documentation](http://msdn.microsoft.com/en-us/library/system.random.aspx) says that the Random class is based on Donald E. Knuth's subtractive random number generator algorithm. For more information, see D. E. Knuth. "The Art of Computer Programming, volume 2: Seminumerical Algorithms". Addison-Wesley, Reading, MA, second edition, 1981.
>
> The problem was discovered when I use .NET Reflector to see what is actually implemented. Knuth was very specific about the lagged Fibonacci sequence coefficient (24 and 55). Somehow Microsoft mistyped the value to be 21 (this.inextp = 0x15 in public Random(int Seed) in source code),  in place of 31 (=55-24). Due to this mistype, the random number generated no longer has the guanrantee on the period to be longer than (2^55-1). The Knuth version of lags (24, 55) has been tested by many people about the property of the generated numbers but not the one created by Microsoft.
>
> It is very easy to identify the typo. It seems that Microsoft just copied the routine (ran3) from Numerical Recipes v.2 page. 283 verbatim (even the variable names are copied), not from Knuth book which has a complete different way of initializing.   You see that Numerical Recipe version uses 31 correctly, not 21 used by Microsoft.

Looking at the [sources](https://github.com/dotnet/coreclr/blob/633f4ed3231fd3977976fa974d13038b338be325/src/mscorlib/shared/System/Random.cs#L90), it seems it's still the case today.

I'm not sure what the implication of our deviation is. If we wanted to fix it, we'd likely would have to quirk the implementation on .NET Framework. For .NET Core, we can debate the merits of quirking, but it's likely we break customers with changing the seed.

Thoughts?

**Comment by GrabYourPitchforks** [comment]
@terrajobst @stephentoub Are we still entertaining these ideas? We know that people depend on the current behavior of `Random` (predictable output given a specific seed). The implementation is flawed, yes, but the biases introduced shouldn't affect real-world consumers in any meaningful way, as this class is meant for _non-cryptographic_ use.

**Comment by stephentoub** [comment]
I don't think so.

**Comment by danmosemsft** [comment]
I agree, This comes up regularly but every time seems to be based on code inspection/theoretical arguments - which are valuable but without widespread feedback from folks who are hitting the limitations in production, I don't think we have enough reason to make a breaking change (nor even add some kind of flag to switch into a new more random mode). I think Random just is how it is, for now.

**Comment by fuglede** [comment]
FWIW, I did not have cryptography in mind when given the example of bias above, and indeed bias is an issue for PRNGs beyond cryptography. In fact, some of the alternatives considered by @colgreen are also predictable (hence not cryptographically secure), yet preferable as they exhibit no apparent bias (and are fast).

Take a couple of examples:

* You're using PRNGs to determine the electrical conductivity of certain materials; a biased PRNG will render your results incorrect (this is a real life example, see Knuth, Sect. 3.2.2).

* You're building an online poker website. Someone figures out that the way cards are dealt is biased. They immediately profit. (In fact, it is customary to use poker as a way to determine if a PRNG is biased or not; see Knuth, Sect. 3.3.2.)

* You're building a bridge and use Monte Carlo methods to determine how to make it strong enough to withstand different kinds of stress scenarios. Biased sampling will pollute the results.

Do you trust that someone will notice something like the 50.34% odd/even bias in practice and report, or rather wait to things start going wrong somewhere? Microsoft may not be liable, and I also appreciate a stable API, but at the very least, make sure to warn about issues you're aware of: That in this particular case not only is `System.Random` not cryptographically secure, without further validation it's worth avoiding in *any* case where high statistical quality is necessary.

**Comment by danmosemsft** [comment]
Perhaps we could use a better PRNG for the new `ulong` API (https://github.com/dotnet/runtime/issues/26741) specifically?

**Comment by colgreen** [comment]
I think @fuglede makes good points. These deficiencies are perhaps minor and insignificant in some use cases, perhaps most, but there will be scenarios - such as in scientific or machine learning tools - where these biases could introduce a certain type of subtle error that leads to bad effects overall.

On that basis, at the very least the documentation should be calling these problems out at a high level at least, and perhaps if the desire to leave System.Random alone is strong, there could be a new improved alternative (e.g. System.Specialized.Random) that the docs refer to as a better choice.

If you don't call out the problems then I suspect these github issues/tickets will keep popping up indefintely. At least if there's an alternative you can just point them at that and close the ticket :)

To my mind the thinking that System.Random "is what it is" is an appropriate and reasonable approach for the .NET Framework, whereas for dotnet core I sense there's more of a progressive, forward looking community spirit, and willingness to change things that aren't quite right.

There is the very real possibility of people assuming System.Random is good quality overall, because it's *System*.*Random*, and look at all of the other improvements that are going on, etc. 

I'll throw more thought out there - performance. I, like many, enjoy reading the dotnet core perfornance blog posts by @stephentoub , and System.Random could be another class with a 'chunky' performance improvement - e.g. the integer routines are all dependent on the Sample() method that works with a double-precisio float.

Bottom line: Is System.Random what you would want it to be going into the future or not? If not, then let's change it! :)  <end sales pitch> :)

**Comment by colgreen** [comment]
Please see these classes I wrote to address all of the known issues:

   https://github.com/colgreen/Redzen/tree/master/Redzen/Random

Specifically,

    https://github.com/colgreen/Redzen/blob/master/Redzen/Random/Xoshiro256StarStarRandom.cs

And the base class where the general purpose routines live for converting random bits into the various Int32, Int64, doubles and so forth:

    https://github.com/colgreen/Redzen/blob/master/Redzen/Random/RandomSourceBase.cs

The routines are all based on an underlying PRNG that generates 64 bits of entropy per call, thus #26741 trivial.

I could probably talk at length on these routines if anyone wishes to discuss.

**Comment by terrajobst** [comment]
I'm OK with not fixing this issue in `Random`, based on @GrabYourPitchforks's [comment above](https://github.com/dotnet/runtime/issues/23198#issuecomment-666783914).

But should conclude that `Random` is insufficient for modern .NET workloads in machine learning, I'd like us to consider fixing `Random` over introducing a new type. I don't doubt that some code will break, but the vast majority of code looking for randomness would work just find an benefit from the improvements.

**Comment by GrabYourPitchforks** [comment]
We can't have it both ways. Either `Random` utilizes a stable algorithm under the covers (as has always been the case in .NET Framework, and has historically been the reason we've not made changes in .NET Core, and has also been the motivating factor for issues like https://github.com/dotnet/runtime/issues/27429 and https://github.com/dotnet/runtime/issues/18996#issuecomment-254793555 and https://github.com/dotnet/runtime/issues/6203#issuecomment-256898208). Or we're free to improve the algorithm (which is hinted at [on MSDN](https://docs.microsoft.com/en-us/dotnet/api/system.random#instantiating-the-random-number-generator), but in a very loose fashion). But we can't exist in both worlds concurrently.

So... which world do we want to be in? :)

**Comment by GrabYourPitchforks** [comment]
FWIW even though MSDN says the algorithm may change from version to version, I don't think this is something we have ever seriously pursued before. That would somewhat defeat the purpose of using seeds in the first place, as it means you're not even guaranteed to be able to replicate results between different invocations of the same process.

**Comment by peteroupc** [comment]
I observe that even though `System.Random`doesn't allow retrieving the seed it uses _directly_, its state could be retrieved _indirectly_ through serialization, since `System.Random` has the SerializableAttribute.  If it weren't for SerializableAttribute, then a `System.Random` instance created through its no-argument constructor could theoretically use a different PRNG algorithm than the one it uses through its `int` constructor, while preserving backward compatibility.

Also note, in my Random Number Generator Recommendations for Applications, I have the following statement (under "Implementing New RNG APIs"):

> If the API provides a PRNG that an application can seed for reproducible "randomness", it should document that PRNG and any methods the API provides that use that PRNG (such as shuffling and Gaussian number generation), and should not change that PRNG or those methods in a way that would change the "random" numbers they deliver for a given seed.

**Comment by danmosemsft** [comment]
> System.Random instance created through its no-argument constructor could theoretically use a different PRNG algorithm than the one it uses through its int constructor, while preserving backward compatibility

That might be the way forward. The parameter less ctor is the commonly used one per https://apisof.net/catalog/System.Random..ctor(Int32)

**Comment by colgreen** [comment]
How about:
  
    cstr() - legacy PRNG
    cstr(int seed) - legacy PRNG
    cstr(RandomAlgorithm prng) - A specified PRNG
    cstr(RandomAlgorithm prng, int seed) - A specified PRNG, with a given seed

And:

    enum RandomAlgorithm 
    {
        Legacy,  // or MersenneTwister, or LegacyMersenneTwister
        Foo,     // E.g. xoshiro256**
        Bar,
        // etc.
    }

Where on day one there may be only a single new PRNG in the enum, but we have the freedome in the future to add new PRNGs as necessary, e.g. as and when problems are reported with the FooPrng.

**Comment by GrabYourPitchforks** [comment]
> That might be the way forward.

I don't think so. The arguments given in this thread (ML scenarios, etc.) revolve around being able to reproduce the random sequence. The `Random` class itself might not be serializable in .NET Core, but the way you'd make this work is to serialize _just the seed_, then to have all applications instantiate a random instance from the persisted seed. This very strongly implies that the desired behavior from this thread is _both_ to use a different algorithm _and_ to have a way to reproduce the output sequence, which means we can't key the improved behavior off of the parameterless ctor.

I further don't think we should have an enum value that controls the algorithm. The two immediate concerns that come to mind are: (a) how does the caller determine which algorithm to use?; and (b) what's the bar for determining which algorithms to include in-box? For most applications we say that if you need a specialized algorithm or data structure that we don't provide in-box, pull in an existing third-party implementation. Or, for frameworks that require some functionality, the frameworks can provide the implementation. In the case of ML, this means that the ML package can provide `Random`-derived types for all the different random number generators that are commonly used by ML scenarios.

**Comment by peteroupc** [comment]
If the intent is that the PRNG algorithm used by `System.Random` will never change, even in future versions of .NET (and perhaps even in the case when a `System.Random` instance is initialized through its no-argument constructor), then the `System.Random` class should stop stating that the algorithm is subject to changes and should document the current algorithm in full, flaws and warts and all, just as Java does with `java.util.Random`.

**Comment by Clockwork-Muse** [comment]
> For most applications we say that if you need a specialized algorithm or data structure that we don't provide in-box, pull in an existing third-party implementation. 

In light of this, one wonders if we should stop providing all the random "helpers" (as in, `NextInt`, etc), and just provide random bytes from the OS, expecting third party libraries to handle this sort of thing.

This is one of the times I wish we'd made packages for almost everything in the standard library, since that would allow us to "fix" this easier.

**Comment by terrajobst** [comment]
> We can't have it both ways. Either `Random` utilizes a stable algorithm under the covers (as has always been the case in .NET Framework, and has historically been the reason we've not made changes in .NET Core, and has also been the motivating factor for issues like #27429 and [#18996 (comment)](https://github.com/dotnet/runtime/issues/18996#issuecomment-254793555) and [#6203 (comment)](https://github.com/dotnet/runtime/issues/6203#issuecomment-256898208)). Or we're free to improve the algorithm (which is hinted at [on MSDN](https://docs.microsoft.com/en-us/dotnet/api/system.random#instantiating-the-random-number-generator), but in a very loose fashion). But we can't exist in both worlds concurrently.
> 
> So... which world do we want to be in? :)

In a world of in-place updates changing the algorithm isn't doable, but that's not the world we live in with .NET Core anymore, so I'd argue it's false dichotomy :-)

We could decide to introduce a new static factory method on the same type or we could decide to make a breaking change between major releases (potentially with an app compat switch). These are all better outcomes than a new type.

**Comment by GrabYourPitchforks** [comment]
> We could decide to introduce a new constructor/static factory method on the same type or we could decide to make a breaking change between major releases (potentially with an app compat switch). These are all better outcomes than a new type.

I don't think this puts us in a better world. This discounts that libraries might be using the `Random` type and might be expecting it to have a given algorithm. An app-wide switch isn't workable for this scenario since it removes the ability of the individual callers to have control over their logic.

**Comment by GrabYourPitchforks** [comment]
FWIW here's where I see things landing:

1. If the caller currently provides a seed, they _require_ consistency with previous versions. There's quite literally no reason for them to have provided a seed other than to have reproducible results. Us changing the algorithm out from under them would break these scenarios.

2. If the caller doesn't provide a seed, no back-compat is needed. By definition they didn't require reproducible output, nor did they rely on any specific algorithm.

3. If the caller needs high-quality random data, their scenario should mandate a tolerance that they're willing to accept. This implies that the scenario itself will dictate what algorithm is to be used, or it's up to the individual caller to ensure that they're using an algorithm that fulfills those requirements. In neither case is a general "we give you some random data but we don't tell you what the algorithm is, and BTW the algorithm is subject to change" API sufficient for this purpose.

My argument hinges on that it's a bad idea for us to change (1) and silently break consumers; and that for (3) callers should be calling specific APIs which are contracted to fulfill their specific requirements, but it's not up to the Framework to provide such algorithms in-box.

**Comment by colgreen** [comment]
I broadly agree with @GrabYourPitchforks, but bear in mind that the PRNG is only part of the problem. There are other issues related to how the various methods (Next, NextDouble...) convert the random bits into the required return types.

It would still be necessary to document well what is going on.

Another quirk of System.Random (that I once encountered causing a problem in the wild), is that the parameterless constructor uses the system clock as the basis for the seed, with the effect that if you construct multiple Random(s) in quick succession, they can all have the same seed! In the Redzen.Random classes I used another PRNG as the source of seed values, which itself is seeded by RNGCryptoServiceProvider - which I believe is seeded by system entropy. I also uses a low locking strategy to ensure this would scale reasonably well if anyone every did want to construct instances at a high rate (which is questionable, but real world code gets messy sometimes). See:

https://github.com/colgreen/Redzen/blob/master/Redzen/Random/DefaultRandomSeedSource.cs

**Comment by danmosemsft** [comment]
> Another quirk of System.Random (that I once encountered causing a problem in the wild), is that the parameterless constructor uses the system clock as the basis for the seed, with the effect that if you construct multiple Random(s) in quick succession, they can all have the same seed!

This is true in .NET Framework, but not .NET Core. The latter uses a private Random based on OS entropy to initialize a per-thread Random that generates seeds. I actually ran into this a few days ago, and .NET Framework was having exactly this problem, but .NET Core did not.

**Comment by peteroupc** [comment]
> I broadly agree with @GrabYourPitchforks, but bear in mind that the PRNG is only part of the problem. There are other issues related to how the various methods (Next, NextDouble...) convert the random bits into the required return types.
> 
> It would still be necessary to document well what is going on.
>

In this sense, see again how this is done, for example, in the documentation for `java.util.Random`, especially in the [`nextInt`](https://docs.oracle.com/javase/8/docs/api/java/util/Random.html#nextInt-int-) and [`nextDouble`](https://docs.oracle.com/javase/8/docs/api/java/util/Random.html#nextDouble--) methods.
